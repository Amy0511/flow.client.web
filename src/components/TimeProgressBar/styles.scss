@import "~@boomerang/boomerang-styles/lib/colors";

.c-time-progress-bar {
  height: 5.25rem;
  width: 100%;
  padding: 2rem 1.25rem 1.25rem 1.25rem;
  background-color: $white;
}

.b-time-progress-bar {
  position: relative;
  height: 2rem;
  width: 100%;
  border-radius: 0.5rem;
  background-color: $gray-20;
}

.b-time-progress-bar__start {
  position: absolute;
  font-size: 1rem;
  font-weight: 600;
  top: -1.25rem;
  left: 0;
}

.b-time-progress-bar__finish {
  position: absolute;
  font-size: 1rem;
  font-weight: 600;
  top: -1.25rem;
  right: 0;
}

.b-time-progress-bar__fillers {
  position: relative;
  height: 100%;
}
