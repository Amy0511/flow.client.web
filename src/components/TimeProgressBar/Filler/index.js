import React from "react";
import classnames from "classnames";
import moment from "moment";
import Tooltip from "@boomerang/boomerang-components/lib/Tooltip";
import { ACTIVITY_STATUSES_TO_TEXT } from "Constants/activityStatuses";
import getHumanizedDuration from "@boomerang/boomerang-utilities/lib/getHumanizedDuration";
import "./styles.scss";

const Filler = ({
  id,
  taskId,
  index,
  isFirstTask,
  isLastTask,
  taskName,
  duration,
  finishTime,
  percentOfTotal,
  updateActiveNode,
  status
}) => {
  const colors = ["#57d785", "#95c4f3", "#e3bc13", "#fc835c", "#f7aac3"];
  const skippedColor = "#6f7878";
  const styles = {
    backgroundColor: status === "skipped" ? skippedColor : colors[index],
    width: `${percentOfTotal}%`
  };

  const finishTimeFormatted = moment.unix(finishTime / 1000).format("hh:mm:ss a", { trim: false });
  return (
    <>
      <div
        onClick={() => updateActiveNode(taskId)}
        data-tip
        data-for={id}
        className={classnames("b-time-progress-bar-filler", { "--first": isFirstTask, "--last": isLastTask })}
        style={styles}
      >
        {percentOfTotal > 5 && <div className="b-time-progress-bar-filler__time">{finishTimeFormatted}</div>}
      </div>

      <Tooltip id={id} theme="bmrg-white" place="bottom">
        <div className="c-time-progress-bar-filler-info">
          <div className="b-time-progress-bar-filler-info">
            <label className="b-time-progress-bar-filler-info__label">Task name:</label>
            <label className="b-time-progress-bar-filler-info__label">Duration:</label>
            <label className="b-time-progress-bar-filler-info__label">Status:</label>
          </div>
          <div className="b-time-progress-bar-filler-info">
            <label className="b-time-progress-bar-filler-info__value">{taskName}</label>
            <label className="b-time-progress-bar-filler-info__value">
              {getHumanizedDuration(parseInt(duration / 1000, 10))}
            </label>
            <label className="b-time-progress-bar-filler-info__value">{ACTIVITY_STATUSES_TO_TEXT[status]}</label>
          </div>
        </div>
      </Tooltip>
    </>
  );
};

export default Filler;
