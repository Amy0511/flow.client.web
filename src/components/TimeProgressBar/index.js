import React from "react";
import PropTypes from "prop-types";
import orderBy from "lodash/orderBy";
import Filler from "./Filler";
import "./styles.scss";

const TimeProgressBar = ({ tasks, updateActiveNode }) => {
  const steps = orderBy(tasks.steps, ["order"], ["asc"]);
  const totalDuration = tasks.duration;
  const allCompleted = !steps.filter(step => step.flowTaskStatus !== "completed").length;

  return (
    <div className="c-time-progress-bar">
      <div className="b-time-progress-bar">
        <div className="b-time-progress-bar__start">Start</div>
        <div className="b-time-progress-bar__finish">Finish</div>
        <div className="b-time-progress-bar__fillers">
          {steps.map((step, index) => (
            <Filler
              key={step.taskId}
              taskId={step.taskId}
              index={index}
              taskName={step.taskName}
              finishPosition={
                allCompleted
                  ? 100
                  : Math.round(((step.startTime - tasks.creationDate + step.duration) / totalDuration) * 100)
              }
              totalDuration={totalDuration}
              updateActiveNode={updateActiveNode}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

TimeProgressBar.propTypes = {
  tasks: PropTypes.object.isRequired,
  updateActiveNode: PropTypes.func
};

export default TimeProgressBar;
