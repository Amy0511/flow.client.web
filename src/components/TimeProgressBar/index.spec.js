import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import TimeProgressBar from "./index";

const mockfn = jest.fn();

const tasks = {
  creationDate: 1540912389131,
  duration: 7178,
  id: "5bd8750446e4bf790fe60aef",
  status: "completed",
  workflowId: "5bd6fdf65a5df954ad47bdbe",
  workflowRevisionid: "5bd85a2146e4bf444d97417d",
  steps: [
    {
      activityId: "5bd8750446e4bf790fe60aef",
      duration: 2178,
      flowTaskStatus: "inProgress",
      id: "5bd8750546e4bf790fe60af0",
      order: 0,
      startTime: 1540912389131,
      taskId: "5006c70f-c7f0-4679-940c-82ba89755599",
      taskName: "task1"
    },
    {
      activityId: "5bd8750446e4bf790fe60aef",
      duration: 3000,
      flowTaskStatus: "notstarted",
      id: "5bd8750546e4bf790fe60af1",
      order: 1,
      startTime: 1540912391309,
      taskId: "d64fb25d-932a-4894-9087-ded1b63eb672",
      taskName: "task2"
    }
  ]
};

describe("TimeProgressBar --- Snapshot", () => {
  it("Capturing Snapshot of TimeProgressBar", () => {
    const renderedValue = renderer.create(<TimeProgressBar tasks={tasks} updateActiveNode={mockfn} />);
    expect(renderedValue).toMatchSnapshot();
  });
});

describe("TimeProgressBar --- Shallow render", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<TimeProgressBar tasks={tasks} updateActiveNode={mockfn} />);
  });

  it("Render the DUMB component", () => {
    expect(wrapper.length).toEqual(1);
  });
});
