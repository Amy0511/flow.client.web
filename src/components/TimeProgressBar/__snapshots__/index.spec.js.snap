// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeProgressBar --- Snapshot Capturing Snapshot of TimeProgressBar 1`] = `
<div
  className="c-time-progress-bar"
>
  <div
    className="b-time-progress-bar"
  >
    <div
      className="b-time-progress-bar__start"
    >
      Start
    </div>
    <div
      className="b-time-progress-bar__finish"
    >
      Finish
    </div>
    <div
      className="b-time-progress-bar__fillers"
    >
      <div
        className="b-time-progress-bar-filler"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#57d785",
            "width": "30.342713847868485%",
            "zIndex": 20,
          }
        }
      >
        <div
          className="b-time-progress-bar-filler__time"
        >
          00:00:02
        </div>
        <div
          className="b-time-progress-bar-filler__name"
        >
          task1
        </div>
      </div>
      <div
        className="b-time-progress-bar-filler"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#95c4f3",
            "width": "72.13708553914739%",
            "zIndex": 19,
          }
        }
      >
        <div
          className="b-time-progress-bar-filler__time"
        >
          00:00:05
        </div>
        <div
          className="b-time-progress-bar-filler__name"
        >
          task2
        </div>
      </div>
    </div>
  </div>
</div>
`;
